File: javalang/selectionsort/RecursiveSelectionSort.java
package javalang.selectionsort;

import java.util.Arrays;

public class RecursiveSelectionSort {
    public static void sort(int[] array) {
        sort(array, 0, array.length - 1);
    }

    public static void sort(int[] array, int l, int r) {
        if (l < r) {
            int currentMin = array[l];
            int currentMinIndex = l;
            for (int i = l + 1; i <= r; i++) {
                if (array[i] < currentMin) {
                    currentMin = array[i];
                    currentMinIndex = i;
                }
            }
            array[currentMinIndex] = array[l];
            array[l] = currentMin;

            sort(array, l + 1, r);
        }
    }

    public static void main(String[] args) {
        System.out.println("testing recursive selection sort in java...");
        int[] unsortedArray = { 9, 5, 7, 1, 6, 2, 3, 8, 4 };

        System.out.println("unsorted array: " + Arrays.toString(unsortedArray));
        RecursiveSelectionSort.sort(unsortedArray);
        System.out.println("sorted array: " + Arrays.toString(unsortedArray));
    }
}

File: c/recursive_selection_sort.c
#include <stdio.h>

void recursive_selection_sort(int *array, int size, int l, int r);

int main() {
    printf("testing recursive selection sort in c...\n");

    int array[] = { 9, 5, 7, 1, 6, 2, 3, 8, 4 };
    int size = sizeof(array) / sizeof(int);

    printf("unsorted array -> [ %d", array[0]);
    for (int i = 1; i < size; i++) {
        printf(", %d", array[i]);
    }
    printf(" ]\n");

    recursive_selection_sort(array, size, 0, size - 1);

    printf("sorted array -> [ %d", array[0]);
    for (int i = 1; i < size; i++) {
        printf(", %d", array[i]);
    }
    printf(" ]\n");

    return 0;
}

void recursive_selection_sort(int *array, int size, int l, int r) {
    if (l < r) {
        int currentMin = array[l];
        int currentMinIndex = l;
        for (int i = l + 1; i < size; i++) {
            if (currentMin > array[i]) {
                currentMin = array[i];
                currentMinIndex = i;
            }
        }
        array[currentMinIndex] = array[l];
        array[l] = currentMin;
        recursive_selection_sort(array, size, l + 1, r);
    }
}